darts/models/forecasting/deeptime.py

```diff
@@ -0,0 +1,661 @@
+            Determines the number of neurons that make up each hidden fully connected layer of the INR module.
+            If a list is passed, it must have a length equal to `num_layers`. If an integer is passed,
+            every layers will have the same width.
+        n_fourier_feats
+            Number of Fourier components to sample to represent to time-serie in the frequency domain
+        scales
+            Scaling factors applied to the normal distribution sampled for Fourier components' magnitude
+        dropout
+            The dropout probability to be used in fully connected layers (default=0). This is compatible with
+            Monte Carlo dropout at inference time for model uncertainty estimation (enabled with
+            ``mc_dropout=True`` at prediction time).
+        activation
+            The activation function of encoder/decoder intermediate layer (default='ReLU').
+            Supported activations: ['ReLU','RReLU', 'PReLU', 'Softplus', 'Tanh', 'SELU', 'LeakyReLU',  'Sigmoid']
+        **kwargs
+            Optional arguments to initialize the pytorch_lightning.Module, pytorch_lightning.Trainer, and
+            Darts' :class:`TorchForecastingModel`.
+        """
+        super().__init__(**kwargs)
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        # TODO: develop support for nr_params functionality
+        self.nr_params = nr_params
+        self.inr_num_layers = inr_num_layers
+        self.inr_layers_width = inr_layers_width
+        self.n_fourier_feats = n_fourier_feats
+        self.scales = scales
+
+        self.dropout = dropout
+        self.activation = activation
+
+        self.inr = INR(
+            input_dim=self.input_dim + 1,
+            num_layers=self.inr_num_layers,
+            hidden_layers_width=self.inr_layers_width,
+            n_fourier_feats=self.n_fourier_feats,
+            scales=self.scales,
+            dropout=self.dropout,
+            activation=self.activation,
+        )
+
+        self.adaptive_weights = RidgeRegressor()
+
+    def forward(self, x_in: Tensor) -> Tensor:
+        x, _ = x_in  # x_in: (past_target|past_covariate, static_covariates)
+
+        batch_size, _, _ = x.shape  # x: (batch_size, in_len, in_dim)
+        coords = self.get_coords(self.input_chunk_length, self.output_chunk_length)
+
+        time_reprs = self.inr(coords)
+        time_reprs = time_reprs.repeat(batch_size, 1, self.input_dim)
+
+        lookback_reprs = time_reprs[:, : -self.output_chunk_length]
+        horizon_reprs = time_reprs[:, -self.output_chunk_length :]
+        # learn weights from the lookback
+        w, b = self.adaptive_weights(lookback_reprs, x)
+        # apply weights to the horizon
+        y = torch.einsum("... d o, ... t d -> ... t o", [w, horizon_reprs]) + b
+
+        # taken from nbeats, could be simplified
+        y = y.view(
+            y.shape[0], self.output_chunk_length, self.input_dim, self.nr_params
```

Comment:
Reshaping to `nr_params` as last dim will likely fail when `nr_params > 1` (e.g. if you set `likelihood=GaussianLikelihood()`. Would there be a natural way to have multiple final outputs of the network? It can perhaps a matter of setting `output_dim * nr_params` instead of `output_dim` somewhere earlier.

Reply:
I can see two possible approaches to increase dimension of the output:
1. One RidgeRegressor per nr_params is applied to the output of the INR network, each solving the equation AX = B_i where each B_i correspond to a nr_params (same encoding of the time axis A for all the nr_params)
2. the INR network can generate one time representation per `nr_params`, by multiplying the last layer by `nr_params` as you suggested. Then, either the same RidgeRegressor is applied to each X matrix or the option 1 can be reused.

By same RidgeRegressor, I mean that the lambda parameter (regularization, learned during the training) is identical. I will start experimenting with option 2 combined with a unique RidgeRegressor and remain open to suggestions.